rules/prefer-array-includes.js

```diff
@@ -0,0 +1,162 @@
+'use strict';
+const {hasSideEffect, isParenthesized, findVariable} = require('eslint-utils');
+const getDocumentationUrl = require('./utils/get-documentation-url');
+const methodSelector = require('./utils/method-selector');
+const getVariableIdentifiers = require('./utils/get-variable-identifiers');
+const getReferences = require('./utils/get-references');
+
+const MESSAGE_ID_SOME = 'some';
+const MESSAGE_ID_REPLACE = 'replaceSome';
+const messages = {
+	[MESSAGE_ID_SOME]: 'Use `.includes()` instead of `.some()` when looking for a well known item.',
+	[MESSAGE_ID_REPLACE]: 'Replace `.some()` with `.includes()`.'
+};
+
+const getBinaryExpressionSelector = path => [
+	`[${path}.type="BinaryExpression"]`,
+	`[${path}.operator="==="]`,
+	`:matches([${path}.left.type="Identifier"], [${path}.right.type="Identifier"])`
+].join('');
+const getFunctionSelector = path => [
+	`[${path}.generator=false]`,
+	`[${path}.async=false]`,
+	`[${path}.params.length=1]`,
+	`[${path}.params.0.type="Identifier"]`
+].join('');
+const selector = [
+	methodSelector({
+		name: 'some',
+		length: 1
+	}),
+	`:matches(${
+		[
+			// Matches `foo.some(bar => bar === baz)`
+			[
+				'[arguments.0.type="ArrowFunctionExpression"]',
+				getFunctionSelector('arguments.0'),
+				getBinaryExpressionSelector('arguments.0.body')
+			].join(''),
+			// Matches `foo.some(bar => {return bar === baz})`
+			// Matches `foo.some(function (bar) {return bar === baz})`
+			[
+				':matches([arguments.0.type="ArrowFunctionExpression"], [arguments.0.type="FunctionExpression"])',
+				getFunctionSelector('arguments.0'),
+				'[arguments.0.body.type="BlockStatement"]',
+				'[arguments.0.body.body.length=1]',
+				'[arguments.0.body.body.0.type="ReturnStatement"]',
+				getBinaryExpressionSelector('arguments.0.body.body.0.argument')
+			].join('')
+		].join(', ')
+	})`
+].join('');
+
+const isIdentifierNamed = ({type, name}, expectName) => type === 'Identifier' && name === expectName;
+
+function isVariablesInCallbackUsed(scopeManager, callback, parameterInBinaryExpression) {
+	const scope = scopeManager.acquire(callback);
+
+	// `parameter` is used on somewhere else
+	const [parameter] = callback.params;
+	if (
+		getVariableIdentifiers(findVariable(scope, parameter))
+			.some(identifier => identifier !== parameter && identifier !== parameterInBinaryExpression)
+	) {
+		return true;
+	}
+
+	// `call` is done
+	if (getReferences(scope).some(reference => reference.identifier.name === 'call' && !reference.resolved)) {
```

Comment:
And I don't understand this, this line is new, why `call` special?

Reply:
```JavaScript
const isFound = foo.some(x => x === call());
```

This call has "side effects".

Reply:
So it's not true? Or it's not a good solution?

Reply:
I don't understand what are you doing, this one should have exactly same logic as `prefer-array-index-of`.

And please read other comments!

Reply:
> > However, if the expression you are looking for has side effects, relies on element related to the function (the name of the function, its arguments, etc.), the case is still valid.
> 
> This part seems not right, the searching value has "side effects" not valid.
> 
> ```js
> const index = foo.findIndex(x => x === call());
> ```
> 
> This `call` has "side effects", but it will still reported without "auto-fix".

I'm just relying on what you have said on `prefer-array-index-of` rule. If I understand well what you have said, `call()` has side effects, so it should be valid. Here, I'm trying to handle this. I know `prefer-array-index-of` rule has been merged without it but it might be taken into account.

Actually if it's not the case or if you want to remove it, I will just remove this part.

> And please read other comments!

I haven't answered because I haven't any answer yet. I should have added acknowledges.

Reply:
You misunderstood, that case is "invalid", but can't auto-fix. Of cause it's handled https://github.com/sindresorhus/eslint-plugin-unicorn/blob/df09d31511825ba383efa466e317f735e35cb45e/test/prefer-array-index-of.js#L105-L108

`call` is just an example, you can't just ignore that name, it can be any function.